analytic_engine/src/instance/wal_replayer.rs

```diff
@@ -0,0 +1,596 @@
+use wal::{
+    log_batch::LogEntry,
+    manager::{
+        ReadBoundary, ReadContext, ReadRequest, RegionId, ScanContext, ScanRequest, WalManagerRef,
+    },
+};
+
+use crate::{
+    instance::{
+        self,
+        engine::{ReplayWalWithCause, Result},
+        flush_compaction::{Flusher, TableFlushOptions},
+        serial_executor::TableOpSerialExecutor,
+        write::MemTableWriter,
+    },
+    payload::{ReadPayload, WalDecoder},
+    table::data::TableDataRef,
+};
+
+/// Wal replayer supporting both table based and region based
+// TODO: limit the memory usage in `RegionBased` mode.
+pub struct WalReplayer<'a> {
+    context: ReplayContext,
+    mode: ReplayMode,
+    table_datas: &'a [TableDataRef],
+}
+
+impl<'a> WalReplayer<'a> {
+    pub fn new(
+        table_datas: &'a [TableDataRef],
+        shard_id: ShardId,
+        wal_manager: WalManagerRef,
+        wal_replay_batch_size: usize,
+        flusher: Flusher,
+        max_retry_flush_limit: usize,
+        replay_mode: ReplayMode,
+    ) -> Self {
+        let context = ReplayContext {
+            shard_id,
+            wal_manager,
+            wal_replay_batch_size,
+            flusher,
+            max_retry_flush_limit,
+        };
+
+        Self {
+            mode: replay_mode,
+            context,
+            table_datas,
+        }
+    }
+
+    fn build_core(mode: ReplayMode) -> Box<dyn ReplayCore> {
+        info!("Replay wal in mode:{mode:?}");
+
+        match mode {
+            ReplayMode::RegionBased => Box::new(RegionBasedCore),
+            ReplayMode::TableBased => Box::new(TableBasedCore),
+        }
+    }
+
+    pub async fn replay(&mut self) -> Result<HashMap<TableId, Result<()>>> {
+        // Build core according to mode.
+        let core = Self::build_core(self.mode);
+        info!(
+            "Replay wal logs begin, context:{}, tables:{:?}",
+            self.context, self.table_datas
+        );
+        let result = core.replay(&self.context, self.table_datas).await;
+        info!(
+            "Replay wal logs finish, context:{}, tables:{:?}",
+            self.context, self.table_datas,
+        );
+
+        result
+    }
+}
+
+pub struct ReplayContext {
+    pub shard_id: ShardId,
+    pub wal_manager: WalManagerRef,
+    pub wal_replay_batch_size: usize,
+    pub flusher: Flusher,
+    pub max_retry_flush_limit: usize,
+}
+
+impl Display for ReplayContext {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReplayContext")
+            .field("shard_id", &self.shard_id)
+            .field("replay_batch_size", &self.wal_replay_batch_size)
+            .field("max_retry_flush_limit", &self.max_retry_flush_limit)
+            .finish()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ReplayMode {
+    RegionBased,
+    TableBased,
+}
+
+/// Replay core, the abstract of different replay strategies
+#[async_trait]
+trait ReplayCore: Send + Sync + 'static {
```

Comment:
How about rename `ReplayCore` to `Replay`.

Reply:
Suitable to name it using a `verb`?

Reply:
Done.